warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/JexlNodes.java

```diff
@@ -252,4 +252,31 @@ public static boolean removeFromParent(JexlNode parent, JexlNode childToRemove)
     public static boolean isNotChildless(JexlNode node) {
         return node != null && node.jjtGetNumChildren() > 0;
     }
+    
+    /**
+     * Ascends the Jexl tree searching for a negation. Search defaults to four nodes
+     *
+     * @param node
+     * @return
+     */
+    public static boolean isNodeNegated(JexlNode node) {
+        return isNodeNegated(node, 4);
+    }
+    
+    /**
+     * Ascends the JexlTree searching for negations according to the specified maximum number of
+     *
+     * @param node
+     * @return
+     */
+    public static boolean isNodeNegated(JexlNode node, int max) {
```

Comment:
Similar comment here with allowing a 'max' parameter.  This has bit us in the past.  Maybe consider allowing an oldest ancestor to be passed, and stopping there at a known point?  Setting a numerical limit leaves too much to chance IMO.

Reply:
I'll remove this.